src/events.js

```diff
@@ -38,6 +38,36 @@ export function emit( eventName, data ) {
 }
 
 /**
+ * Same as `emit` above. If any callbacks return promises, `emitWithPromise` will
+ * return a promise that waits on all of them.
+ *
+ * @private
+ * @method emitWithPromise
+ * @param {String} eventName
+ * @param {Object} data
+ * @return {Void}
+ */
+export function emitWithPromise( eventName, data ) {
+	if ( objectType( eventName ) !== "string" ) {
+		throw new TypeError( "eventName must be a string when emitting an event" );
+	}
+
+	// Clone the callbacks in case one of them registers a new callback
+	const originalCallbacks = LISTENERS[ eventName ];
+	const callbacks = originalCallbacks ? [ ...originalCallbacks ] : [];
+
+	let promises = undefined;
```

Comment:
I think we can simplify this by doing:

```js
const promises = callbacks.map(callback => Promise.resolve(callbacks[i](data)));
return Promise.all(promises);
```

I think it'll be cleaner if we always treat them as Promises, especially when we account for my other comment.